runtime/rococo/src/propose_parachain.rs

```diff
@@ -0,0 +1,383 @@
+		/// All the proposals.
+		Proposals: map hasher(twox_64_concat) ParaId => Option<Proposal<T::AccountId, T::ValidatorId, BalanceOf<T>>>;
+		/// Proposals that are approved.
+		ApprovedProposals: Vec<ParaId>;
+		/// Proposals that are scheduled at for a fixed session to be applied.
+		ScheduledProposals: map hasher(twox_64_concat) SessionIndex => Vec<ParaId>;
+		/// Information about the registered parachains.
+		ParachainInfo: map hasher(twox_64_concat) ParaId => Option<RegisteredParachainInfo<T::AccountId, T::ValidatorId>>;
+		/// Validators that should be retired, because their Parachain was deregistered.
+		ValidatorsToRetire: Vec<T::ValidatorId>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Config> for enum Call where origin: <T as frame_system::Config>::Origin {
+		type Error = Error<T>;
+
+		/// The maximum name length of a parachain.
+		const MaxNameLength: u32 = T::MaxNameLength::get();
+
+		/// The deposit that will be reserved when proposing a parachain.
+		const ProposeDeposit: BalanceOf<T> = T::ProposeDeposit::get();
+
+		fn deposit_event() = default;
+
+		/// Propose a new parachain
+		///
+		/// This requires:
+		/// - `para_id`: The id of the parachain.
+		/// - `name`: The name of the parachain.
+		/// - `validation_function`: The wasm runtime of the parachain.
+		/// - `initial_head_state`: The genesis state of the parachain.
+		/// - `validators`: Validators that will validate for the relay chain, needs to be at least one.
+		/// - `balance`: The initial balance of the parachain on the relay chain.
+		///
+		/// It will reserve a deposit from the sender account over the lifetime of the chain.
+		#[weight = 1_000_000]
+		fn propose_parachain(
+			origin,
+			para_id: ParaId,
+			name: Vec<u8>,
+			validation_code: ValidationCode,
+			genesis_head: HeadData,
+			validators: Vec<T::ValidatorId>,
+			balance: BalanceOf<T>,
+		) {
+			let who = ensure_signed(origin)?;
+
+			ensure!(name.len() <= T::MaxNameLength::get() as usize, Error::<T>::NameTooLong);
+			ensure!(validators.len() > 0, Error::<T>::AtLeastOneValidatorRequired);
+			ensure!(!Proposals::<T>::contains_key(&para_id), Error::<T>::ParachainIdAlreadyProposed);
+			ensure!(
+				!runtime_parachains::paras::Module::<T>::parachains().contains(&para_id),
+				Error::<T>::ParachainIdAlreadyTaken,
+			);
+			ensure!(
+				!runtime_parachains::paras::Module::<T>::upcoming_paras().contains(&para_id),
+				Error::<T>::ParachainIdAlreadyTaken,
+			);
+			ensure!(validation_code.0.starts_with(runtime_common::WASM_MAGIC), Error::<T>::DefinitelyNotWasm);
+
+			let active_validators = Session::<T>::validators();
+			ensure!(
+				validators.iter().all(|v| !active_validators.contains(v)),
+				Error::<T>::ValidatorAlreadyRegistered,
+			);
+			Proposals::<T>::iter().try_for_each(|(_, prop)|
+				if validators.iter().all(|v| !prop.validators.contains(v)) {
+					Ok(())
+				} else {
+					Err(Error::<T>::ValidatorAlreadyRegistered)
+				}
+			)?;
```

Comment:
This is pretty aggressive, if I were writing a production pallet, i would try to catch this error on approval. If i approve a pallet, an the validator is already registered in the system, then it would fail. This should prevent any duplicate registrations from getting into the pallet.

Reply:
Versus looking through all proposals proactively, which really has no bounds as far as I can tell.

Reply:
I see what you mean. However, even on approval we should still iterate over all approved proposals to search for validators.

I hope that this pallet never gets used for anything than for this test network, so I think it is okay to be "aggressive" :see_no_evil: